当ch实例为集群时，新增数据库不显示集群选项

使用最新版clickvisual-v0.4.0-rc3

我安装了ch 3分片2副本的集群，当我创建实例为集群时，新增数据库不显示集群选项

![image](https://user-images.githubusercontent.com/41657218/185772047-662e9822-08e3-4441-a008-939ff781484b.png)


![image](https://user-images.githubusercontent.com/41657218/185772053-ff099b4f-2853-4599-b327-d88926f94f37.png)

日志：
```
2022/08/21 14:58:11 /home/runner/go/pkg/mod/gorm.io/driver/mysql@v1.3.5/migrator.go:181
[0.809ms] [rows:-] SELECT column_name, column_default, is_nullable = 'YES', data_type, character_maximum_length, column_type, column_key, extra, column_comment, numeric_precision, numeric_scale , datetime_precision FROM information_schema.columns WHERE table_schema = 'clickvisual' AND table_name = 'cv_pms_casbin_rule' ORDER BY ORDINAL_POSITION

2022/08/21 14:58:11 /home/runner/work/clickvisual/clickvisual/api/internal/service/install/install.go:127 Error 1062: Duplicate entry '1' for key 'PRIMARY'
[0.322ms] [rows:0] INSERT INTO `cv_pms_casbin_rule` VALUES (1, 'p', 'role__root', '*', '*', '*', '', '', '','');

2022/08/21 14:58:11 /home/runner/work/clickvisual/clickvisual/api/internal/service/install/install.go:128 Error 1062: Duplicate entry '2' for key 'PRIMARY'
[0.245ms] [rows:0] INSERT INTO `cv_pms_casbin_rule` VALUES (2, 'g3', 'user__1', 'role__root', '', '', '', '', '', '');
```

另外，我将架构设置为：vector(任意个)+kafka集群+负载均衡(3个clickvistal)+clickhouse集群

--- a/ui/src/components/BreadCrumbs/index.less
+++ b/ui/src/components/BreadCrumbs/index.less
@@ -1,6 +1,6 @@
 .nameSpan {
     display: inline-block;
-    width: 115px;
+    max-width: 115px;
     overflow: hidden;
     text-overflow: ellipsis;
     white-space:nowrap;

--- a/ui/src/components/BreadCrumbs/index.tsx
+++ b/ui/src/components/BreadCrumbs/index.tsx
@@ -2,7 +2,8 @@ import styles from "./index.less";
 import { Tooltip } from "antd";
 import { QUERY_PATH } from "@/config/config";
 import { logLibraryInfoType } from "@/components/BreadCrumbs/type";
-import { CSSProperties } from "react";
+import { CSSProperties, useEffect, useState } from "react";
+import { getTextWith } from "@/utils/textWith";
 interface BreadCrumbsProps {
   logLibraryInfo: logLibraryInfoType;
   style?: CSSProperties;
@@ -11,10 +12,26 @@ interface BreadCrumbsProps {
 
 const BreadCrumbs = (props: BreadCrumbsProps) => {
   const { logLibraryInfo, style, separator } = props;
+  const [tableWidth, setTableWidth] = useState<number>(115);
 
   const getGoToQueryPagePathByTid = (tid?: number) => {
     return `${QUERY_PATH}?tid=${tid}`;
   };
+
+  useEffect(() => {
+    // 将实例、数据库的文字放到canvas里面，获取文本的长度，然后计算table字段最多还能占据的长度
+    const instanceWidth =
+      (getTextWith(logLibraryInfo.instanceName) || 0) > 115
+        ? 115
+        : getTextWith(logLibraryInfo.instanceName) || 0;
+
+    const databaseWidth =
+      (getTextWith(logLibraryInfo.databaseName) || 0) > 115
+        ? 115
+        : getTextWith(logLibraryInfo.databaseName) || 0;
+
+    setTableWidth(325 - instanceWidth - databaseWidth);
+  }, [logLibraryInfo]);
 
   return (
     <div style={style}>
@@ -52,22 +69,33 @@ const BreadCrumbs = (props: BreadCrumbsProps) => {
         )}
         {logLibraryInfo.databaseName && (
           // <Tooltip title={logLibraryInfo.databaseDesc}>
-          <span className={styles.nameSpan}>
-            &nbsp;{separator || "|"}&nbsp;
-            {logLibraryInfo.databaseName}
-          </span>
+          <>
+            <span className={styles.nameSpan}>
+              &nbsp;{separator || "/"}&nbsp;
+            </span>
+            <span className={styles.nameSpan}>
+              {logLibraryInfo.databaseName}
+            </span>
+          </>
         )}
         {logLibraryInfo.tableName && (
           // <Tooltip title={logLibraryInfo.tableDesc}>
-          <span className={styles.nameSpan}>
-            &nbsp;{separator || "|"}&nbsp;
-            <a
-              href={getGoToQueryPagePathByTid(logLibraryInfo.tid)}
-              target="_blank"
+          <>
+            <span className={styles.nameSpan}>
+              &nbsp;{separator || "/"}&nbsp;
+            </span>
+            <span
+              className={styles.nameSpan}
+              style={{ maxWidth: tableWidth + "px" }}
             >
-              {logLibraryInfo.tableName}
-            </a>
-          </span>
+              <a
+                href={getGoToQueryPagePathByTid(logLibraryInfo.tid)}
+                target="_blank"
+              >
+                {logLibraryInfo.tableName}
+              </a>
+            </span>
+          </>
         )}
       </Tooltip>
     </div>

--- a/ui/src/components/BreadCrumbs/type.ts
+++ b/ui/src/components/BreadCrumbs/type.ts
@@ -1,9 +1,9 @@
 export interface logLibraryInfoType {
-  instanceName?: string;
+  instanceName: string;
   instanceDesc?: string;
-  databaseName?: string;
+  databaseName: string;
   databaseDesc?: string;
-  tableName?: string;
+  tableName: string;
   tableDesc?: string;
   tid?: number;
   did?: number;

--- a/ui/src/models/dataLogs.ts
+++ b/ui/src/models/dataLogs.ts
@@ -90,6 +90,9 @@ const DataLogsModel = () => {
   const cancelTokenLogsRef = useRef<Canceler | null>(null);
   const CancelToken = Request.CancelToken;
 
+  // 最近一次正在加载的tid
+  const [lastLoadingTid, setLastLoadingTid] = useState<number>(0);
+
   const {
     logLibraryCreatedModalVisible,
     logLibraryInfoDrawVisible,
@@ -194,6 +197,10 @@ const DataLogsModel = () => {
 
   const onChangeVisibleIndexModal = (visible: boolean) => {
     setVisibleIndexModal(visible);
+  };
+
+  const onChangeLastLoadingTid = (tid: number) => {
+    setLastLoadingTid(tid);
   };
 
   const onChangeLogPane = (tabPane: PaneType) => {
@@ -360,6 +367,7 @@ const DataLogsModel = () => {
     cancelTokenHighChartsRef.current?.();
     const currentPane = logPanesHelper.logPanes[id.toString()];
     const histogramChecked = currentPane?.histogramChecked ?? true;
+    onChangeLastLoadingTid(id);
     if (!!extra?.isPaging || !!extra?.isOnlyLog || !histogramChecked) {
       const logsRes = await getLogs.run(
         id,
@@ -650,10 +658,12 @@ const DataLogsModel = () => {
     editView,
     isModifyLog,
     doGetViewInfo,
+    lastLoadingTid,
     onChangeViewIsEdit,
     onChangeViewVisibleModal,
     onChangeViewsVisibleDraw,
     onChangeIsModifyLog,
+    onChangeLastLoadingTid,
 
     foldingState,
     onChangeFoldingState,

--- a/ui/src/pages/DataLogs/components/DataSourceMenu/LogLibraryList/LogLibraryItem/index.tsx
+++ b/ui/src/pages/DataLogs/components/DataSourceMenu/LogLibraryList/LogLibraryItem/index.tsx
@@ -56,6 +56,7 @@ const LogLibraryItem = (props: LogLibraryItemProps) => {
     resetLogPaneLogsAndHighCharts,
     onChangeIsModifyLog,
     onChangeCurrentEditLogLibrary,
+    onChangeLastLoadingTid,
   } = useModel("dataLogs");
   const { logPanes, paneKeys, addLogPane, removeLogPane } = logPanesHelper;
 
@@ -113,7 +114,6 @@ const LogLibraryItem = (props: LogLibraryItemProps) => {
   };
 
   const doDeleted = () => {
-    // if (!currentDatabase) return;
     const hideMessage = message.loading(
       {
         content: i18n.formatMessage(
@@ -141,7 +141,6 @@ const LogLibraryItem = (props: LogLibraryItemProps) => {
             },
             3
           );
-          // doGetLogLibraryList();
           onGetList();
           // 不在打开的日志库中
           if (!paneKeys.includes(currentKey)) return;
@@ -300,6 +299,7 @@ const LogLibraryItem = (props: LogLibraryItemProps) => {
           <span
             onClick={() => {
               if (currentLogLibrary?.id === logLibrary.id) return;
+              onChangeLastLoadingTid(logLibrary.id);
               onChangeLogLibrary(logLibrary);
               resetCurrentHighChart();
               onChangePanes();

--- a/ui/src/pages/DataLogs/components/DataSourceMenu/ModalCreatedLogLibrary/index.tsx
+++ b/ui/src/pages/DataLogs/components/DataSourceMenu/ModalCreatedLogLibrary/index.tsx
@@ -59,8 +59,8 @@ const ModalCreatedLogLibrary = (props: { onGetList: any }) => {
               })
             );
             onGetList();
+            onChangeLogLibraryCreatedModalVisible(false);
           }
-          onChangeLogLibraryCreatedModalVisible(false);
         })
         .catch(() => onChangeLogLibraryCreatedModalVisible(false));
     },

--- a/ui/src/pages/DataLogs/components/QueryResult/Content/RawLog/RawLogsOperations/index.tsx
+++ b/ui/src/pages/DataLogs/components/QueryResult/Content/RawLog/RawLogsOperations/index.tsx
@@ -23,7 +23,6 @@ const RawLogsOperations = ({ oldPane }: { oldPane: PaneType | undefined }) => {
 
   const i18n = useIntl();
 
-  // todo: logPanes[currentLogLibrary?.id || 0] 和 oldPane 是否是一样的？
   const performTime = useMemo(() => {
     return logPanes[currentLogLibrary?.id || 0]?.logs?.cost;
   }, [logPanes]);
@@ -36,7 +35,6 @@ const RawLogsOperations = ({ oldPane }: { oldPane: PaneType | undefined }) => {
       {performTime ? (
         <div style={{ flex: 1, textAlign: "right", marginRight: "20px" }}>
           {i18n.formatMessage({ id: "log.perform.time" })}：
-          {/* todo: performTime 和 logPanes[currentLogLibrary?.id || 0]?.logs?.cost 不是一样的吗？*/}
           {logPanes[currentLogLibrary?.id || 0]?.logs?.cost}ms
         </div>
       ) : null}

--- a/ui/src/pages/DataLogs/components/QueryResult/Content/RawLog/index.tsx
+++ b/ui/src/pages/DataLogs/components/QueryResult/Content/RawLog/index.tsx
@@ -9,13 +9,15 @@ import { useIntl } from "umi";
 import ManageIndexModal from "@/pages/DataLogs/components/QueryResult/Content/RawLog/RawLogsIndexes/ManageIndexModal";
 import { useMemo } from "react";
 
-const RawLogContent = () => {
+const RawLogContent = (props: { tid: string }) => {
+  const { tid } = props;
   const {
     currentLogLibrary,
     logsLoading,
     highChartLoading,
     isHiddenHighChart,
     logPanesHelper,
+    lastLoadingTid,
   } = useModel("dataLogs");
   const { logPanes } = logPanesHelper;
 
@@ -26,13 +28,17 @@ const RawLogContent = () => {
 
   const i18n = useIntl();
 
+  console.log(lastLoadingTid, tid);
+
   return (
     <div className={queryResultStyles.content}>
       <RawLogsIndexes tid={currentLogLibrary?.id} />
       <div className={queryResultStyles.queryDetail}>
         {oldPane?.histogramChecked && (
           <Spin
-            spinning={highChartLoading}
+            spinning={
+              lastLoadingTid == parseInt(tid) ? highChartLoading : false
+            }
             tip={i18n.formatMessage({ id: "spin" })}
             wrapperClassName={classNames(
               queryResultStyles.querySpinning,
@@ -45,7 +51,7 @@ const RawLogContent = () => {
           </Spin>
         )}
         <Spin
-          spinning={logsLoading}
+          spinning={lastLoadingTid == parseInt(tid) ? logsLoading : false}
           tip={i18n.formatMessage({ id: "spin" })}
           wrapperClassName={classNames(
             queryResultStyles.querySpinning,

--- a/ui/src/pages/DataLogs/components/QueryResult/index.tsx
+++ b/ui/src/pages/DataLogs/components/QueryResult/index.tsx
@@ -15,8 +15,8 @@ const SharePath = [
   process.env.PUBLIC_PATH + "share/",
 ];
 
-const QueryResult = (props: { tid?: string; activeKey?: string }) => {
-  const { tid, activeKey } = props;
+const QueryResult = (props: { tid: string }) => {
+  const { tid } = props;
   const [usrState] = useUrlState<any>();
   const { statisticalChartsHelper } = useModel("dataLogs");
   const { onSetLocalData } = useLocalStorages();
@@ -39,17 +39,17 @@ const QueryResult = (props: { tid?: string; activeKey?: string }) => {
       }
     };
   }, []);
+
   const content = useMemo(() => {
-    if (activeKey != tid) return <></>;
     switch (activeQueryType) {
       case QueryTypeEnum.LOG:
-        return <RawLogContent />;
+        return <RawLogContent tid={tid} />;
       case QueryTypeEnum.TABLE:
         return <StatisticalTableContent isShare={isShare} />;
       default:
-        return <RawLogContent />;
+        return <RawLogContent tid={tid} />;
     }
-  }, [activeQueryType, usrState, usrState?.mode, activeKey, tid]);
+  }, [activeQueryType, usrState, usrState?.mode]);
 
   return (
     <div

--- a/ui/src/pages/DataLogs/components/RawLogTabs/index.tsx
+++ b/ui/src/pages/DataLogs/components/RawLogTabs/index.tsx
@@ -96,10 +96,9 @@ const RawLogTabs = () => {
             return (
               pane && (
                 <TabPane key={pane.paneId} tab={pane.pane} forceRender>
-                  <QueryResult
-                    tid={pane.paneId}
-                    activeKey={currentLogLibrary?.id.toString()}
-                  />
+                  {pane.paneId === currentLogLibrary?.id.toString() ? (
+                    <QueryResult tid={pane.paneId} />
+                  ) : null}
                 </TabPane>
               )
             );

--- a/ui/src/utils/textWith.ts
+++ b/ui/src/utils/textWith.ts
@@ -1,0 +1,11 @@
+export const getTextWith = (text: string, fontStyle?: any) => {
+  var canvas = document.createElement("canvas");
+  var context = canvas.getContext("2d");
+  if (!context) return;
+  context.font =
+    fontStyle ||
+    "14px -apple-system BlinkMacSystemFont PingFang SC Helvetica Tahoma Arial Microsoft YaHei 微软雅黑 黑体 Heiti sans-serif SimSun 宋体 serif"; // 设置字体样式，当然，也可以在这里给一个默认值
+
+  var dimension = context.measureText(text);
+  return dimension.width;
+};

